### DISCUSSION TITLE
memory saturation problem

### DISCUSSION THREAD

**Discussion by PaoloPerrotta** [pymupdf/PyMuPDF#2621] [Looking for help]
hello, lately we have been using fitz a lot in order to paste some text on pdf files.

yesterday we encountered a problem, when executing the following snippet:

```
import fitz

pdf_document = fitz.open()
page = pdf_document.new_page(width=595.3499755859375, height=841.9500122070312) 

rect = fitz.Rect(412.9554443359375, 268.0219421386719, 429.0594787597656, 344.8525390625)

writer = fitz.TextWriter(page.rect)
writer.fill_textbox(rect=rect.round(),
                    text="& |௯|n|~|n|௯||n",
                    fontsize=2.878290297195706,
                    font=fitz.Font("times"),
                    pos=fitz.Point(rect.tl.x, rect.bl.y - 73.95230662663242),
                    align=0)
writer.write_text(page=page)


pdf_document.save("document.pdf")
pdf_document.close()
```
If you run it locally, you will notice how this operation which normally ends instantly, continues indefinitely and, the local computer memory dramatically increases.

It seems that the sizes of the various objects it has to create are conflicting. Putting `int(font_size)`, which is rounded down, seems to solve the question for this document.

But it is very strange because by experimenting by also putting `font_size = 3 `(which is therefore still greater than 2.87 ...) everything still works :thinking: .

For this case, stripping the `.round()` from the `rect `or `pos` argument also seems to work. However, these modifications cause the pasting of other texts to fail.

is this a bug? or are we making a wrong use of the fill_textbox method?

this happens on `pymupdf = 1.21.1` on both Windows and Unix OS

**Comment by JorjMcKie** [comment]
This is weird! It works for fontsize 3 and fontsize 2. So I need to find the time for debugging this obviously unfortunate fontsize value. Values up to 2.85 work ...

**Reply by PaoloPerrotta** [reply]
Thank you for your answer, if it can help debugging, i found the loop that creates the problem:

this while block here, on line **4408** of the module **utils.py** from fitz
```
n = len(words)
while True:
    line0 = " ".join(words[:n])
    wl = sum(word_lengths[:n]) + space_len * (len(word_lengths[:n]) - 1)
    if wl <= width:
        new_lines.append((line0, wl))
        words = words[n:]
        word_lengths = word_lengths[n:]
        n = len(words)
        line0 = None
    else:
        n -= 1

    if len(words) == 0:
        break
```

it seems to never exit from that while, and in the same time increasing gradually memory consumption, i cannot really undestand why though, probably some rounding in the formula?
As further information i would advice that the fontsize causing the loop changes with the pdf and other params, so it's a really tricky problem i guess

thanks again !
